src/core/internal/hash.d

```diff
@@ -79,14 +79,62 @@ size_t hashOf(T)(auto ref T val, size_t seed = 0) if (!is(T == enum) && __traits
 {
     static if(__traits(isFloating, val))
     {
-        T data = (val != val) ? T.nan : val;
-        auto bytes = toUbyte(data);
-        return bytesHash(bytes.ptr, bytes.length, seed);
+        static if (T.mant_dig == float.mant_dig && T.sizeof == uint.sizeof)
+        {
+            return hashOf(*cast(const uint*) &val, seed);
+        }
+        else static if (T.mant_dig == double.mant_dig && T.sizeof == ulong.sizeof)
+        {
+            return hashOf(*cast(const ulong*) &val, seed);
+        }
+        else
+        {
+            T data = (val != val) ? T.nan : val;
+            auto bytes = toUbyte(data);
+            return bytesHash(bytes.ptr, bytes.length, seed);
+        }
     }
     else
     {
-        auto bytes = toUbyte(val);
-        return bytesHash(bytes.ptr, bytes.length, seed);
+        static if (T.sizeof <= size_t.sizeof && __traits(isIntegral, T))
+        {
+            static if (size_t.sizeof < ulong.sizeof)
+            {
+                //MurmurHash3 32-bit single round
+                enum uint c1 = 0xcc9e2d51;
+                enum uint c2 = 0x1b873593;
+                enum uint c3 = 0xe6546b64;
+                enum uint r1 = 15;
+                enum uint r2 = 13;
+            }
+            else
+            {
+                //Half of MurmurHash3 64-bit single round
+                //(omits second interleaved update)
+                enum ulong c1 = 0x87c37b91114253d5;
+                enum ulong c2 = 0x4cf5ad432745937f;
+                enum ulong c3 = 0x52dce729;
+                enum uint r1 = 31;
+                enum uint r2 = 27;
+            }
+            auto h = c1 * val;
+            h = (h << r1) | (h >>> (typeof(h).sizeof * 8 - r1));
+            h = (h * c2) ^ seed;
+            h = (h << r2) | (h >>> (typeof(h).sizeof * 8 - r2));
```

Comment:
As rotl32 is used by `bytesHash`, it might have made sense to move rotl32 / rotl64 in separate functions.